rfcs/72-editor.md

```diff
@@ -0,0 +1,508 @@
+## Motivation
+
+Unlike most major game engines, Bevy lacks an interactive editor. This lack has
+three major downsides:
+
+1. it increases the barrier to entry for potential new developers;
+2. this forces scene designers to learn to create scenes with code; and
+3. developers must rebuild their project before seeing any changes they make.
+
+While teams can mitigate the last two points by creating project specific
+tools, this does require them to divert resources building those tools that they
+could have spent building the game.
+
+Bevy's plugin-heavy design poses several challenges to making a traditional
+editor. Primarily, unlike other game engines, Bevy does limit developers access
+to the core game loop, similar instead to many game frameworks. While an
+interactive editor could be built on top of Bevy that takes away this freedom;
+doing so violates the spirit of Bevy's plugin-heavy design, thus robbing it of
+Bevy's expressive nature.
+
+**Bevy's editor must acknowledge the duel nature of Bevy as both a game engine
+and a game framework.**
+
+A secondary, smaller challenge, is that Bevy relies on the Cargo build system
+for modularity. An interactive-editor that is core to Bevy must both respect
+this design, and work with the limitations and abilities of Cargo.
+
+## User-facing explanation
+
+The easiest way to get started with Bevy's editor by downloading it with Cargo:
+
+```
+$ cargo install bevy
+```
+
+(Note the use of `install`, not `add`!)
+
+This will add `bevy` to our path:
+
+```
+$ bevy
+Bevy's Command Line Tool
+
+Usage: bevy [OPTIONS] [COMMAND]
+
+...
+```
+
+This tool allows us to create new projects, manage assets, edit projects, and
+even publish games. Note that `bevy` is not meant to replace `cargo`, but to
+work alongside it.
+
+Also, just like Cargo, the Bevy version used in a project does not need to match
+the version of `bevy` you have installed. You can upgrade and downgrade them
+independently. Therefore, we recommend you always use the latest version of the
+`bevy` tool.
+
+Bevy's editor can be used graphically, or as a command line tool. This document
+covers the command line usage, but running `bevy edit` in a project will open
+the graphical view.
+
+**The `bevy` command line interface is not stable. Do not use it in build
+scripts!**
+
+### Default Project Structure
+
+The first entry point for the `bevy` command is `new` and `init`, for creating
+new projects, meant to mimic the commands in cargo:
+
+```
+bevy new [options] <project-name>
+```
+
+As an example, let's create a new project, `flappers`:
+
+```
+$ bevy new flappers
+```
+
+This creates a new `flappers` folder in your current directory, using the
+default template, similar to using `cargo new`. Unlike Cargo, however, a Bevy
+project can start from any number of templates you can get from the Bevy asset
+repository.
+
+Next, let's take a look at the project structure:
+
+```
+flappers/
+    README.md
+    Cargo.toml
+    LICENSE-Apache2.txt
```

Comment:
We should default to dual MIT + Apache.

Reply:
I agree we shouldn't add license files in the template. I can see this turning into complaints when commits accidentally include unwanted licenses.

Reply:
If you make the project creation a bit more interactive, then you could ask the user which kind of license, if any, should be added.
A bit like Rustup asks you which Rust version and modules should be installed.

Reply:
Mmm...perhaps this is a bit unclear on my part. What I imagined was that licensing is separate from the template, but still provided as a (changeable) default. The process for new project construction that I'm imagining is:

1. Fill out template.
2. Add chosen license
3. Add version control
4. Add CI template

This is because the license you choose to release your project under is (imo, usually) a separate concern from which template you use to start your project. And since the cross product of all licenses and templates makes sense, they should be seperate flags.

So something like:

```
bevy new -l mit -l gplv2 -t cardgame my-project
```

I do think we should by default add a license (unless the user explicitly passes something like `-l nolicense`), even if its an 'all rights reserved' statement.

Reply:
It should definitely not add any license by default. That is unacceptable.

Licensing a Bevy project is niche enough that I'd leave it out of `bevy new` entirely.
`cargo new` doesn't provide this functionality either.

Reply:
Cargo doesn't, but these do:

1. Npm (Javascript) does.
2. Raco (Racket) does.
3. Leiningen (Clojure) does.
4. Bevy's existing template on github

Although notably these don't:
1. Sbt (Scala)
2. Pkg (Julia)
3. Godot

Another option is to create a LICENSE file and put text in it saying `<INSERT LICENSE>` as a default.

If we don't want to pick any particular license as a default, we can also just put something like `<Insert License Here>` as a default.

Reply:
I would start without license and later on add a option to include license, because I still feel like the user should opt in for specifying the license, not the tool.

Also the vast majority of Bevy users will make closed source games, which normally is copyright, so no license by default fits the majority of use cases.

Godot, Unity and Unreal doesn't do that, and they should be our benchmark, not package managers, since Bevy is aiming to be a game editor.